firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreMultiDbComponent.java

```diff
@@ -79,10 +80,15 @@ public synchronized void remove(@NonNull String databaseId) {
 
   @Override
   public synchronized void onDeleted(String firebaseAppName, FirebaseOptions options) {
-    // Shuts down all database instances and remove them from registry map when App is deleted.
-    for (Map.Entry<String, FirebaseFirestore> entry : instances.entrySet()) {
-      entry.getValue().terminateInternal();
-      instances.remove(entry.getKey());
+    // Shut down all database instances and remove them from the registry map. To avoid
+    // ConcurrentModificationException, make a copy of the entries instead of using an iterator from
+    // the `instances` map directly.
+    for (Map.Entry<String, FirebaseFirestore> entry : new ArrayList<>(instances.entrySet())) {
+      entry.getValue().terminate();
+      if (instances.containsKey(entry.getKey())) {
```

Comment:
This is just a sanity check to make sure that the `instances` map is in the expected state after the `for` loop completes. If `terminate()` is ever changed to not remove the instance then that would go unnoticed, breaking the contract of this method.

Reply:
`hardAssert` is what you want for this.

Reply:
Done.